idr/marshal1.go

```diff
@@ -0,0 +1,55 @@
+package idr
+
+import (
+	"fmt"
+
+	"github.com/jf-tech/go-corelib/jsons"
+	"github.com/jf-tech/go-corelib/strs"
+)
+
+// j1NodePtrName returns a categorized name for a *Node pointer used in JSONify1.
+func j1NodePtrName(n *Node) *string {
+	if n == nil {
+		return nil
+	}
+	name := func(n *Node) string {
+		if IsXML(n) && XMLSpecificOf(n).NamespacePrefix != "" {
+			return XMLSpecificOf(n).NamespacePrefix + ":" + n.Data
+		}
+		return n.Data
+	}
+	switch n.Type {
+	case DocumentNode:
+		return strs.StrPtr(fmt.Sprintf("(%s)", n.Type))
+	case ElementNode, AttributeNode:
+		return strs.StrPtr(fmt.Sprintf("(%s %s)", n.Type, name(n)))
+	case TextNode:
+		return strs.StrPtr(fmt.Sprintf("(%s '%s')", n.Type, n.Data))
+	default:
+		return strs.StrPtr(fmt.Sprintf("(unknown '%s')", n.Data))
+	}
+}
+
+// j1NodeToInterface converts *Node into an interface{} suitable for json marshaling used in JSONify1.
+func j1NodeToInterface(n *Node) interface{} {
+	m := make(map[string]interface{})
+	m["Parent"] = j1NodePtrName(n.Parent)
+	m["FirstChild"] = j1NodePtrName(n.FirstChild)
+	m["LastChild"] = j1NodePtrName(n.LastChild)
+	m["PrevSibling"] = j1NodePtrName(n.PrevSibling)
+	m["NextSibling"] = j1NodePtrName(n.NextSibling)
+	m["Type"] = n.Type.String()
+	m["Data"] = n.Data
+	m["FormatSpecific"] = n.FormatSpecific
+	var children []interface{}
+	for child := n.FirstChild; child != nil; child = child.NextSibling {
+		children = append(children, j1NodeToInterface(child))
+	}
+	m["Children"] = children
+	return m
+}
+
+// JSONify1 json marshals a *Node verbatim. Mostly used in test for snapshotting.
+func JSONify1(n *Node) string {
```

Comment:
BTW, will you follow this index based naming style in the future? e.g. `marshal1`, `marshal2`, `JSONify1`, `JSONify2` etc.

Reply:
this is just an one off thing. We need two marshaling: 1) full verbatim marshaling for test snapshotting (JSONify1) and 2) the Node -> JSON translated marshaling for used in a) javascript and b) result_type="object".

The b) was something we did after you guys. A feature request came that some schema writer wants to simply copy a certain node directly into output, without any processing (see https://github.com/jf-tech/omniparser/blob/master/handlers/omni/v2/transform/parse.go#L266).

Realizing this feature is essentially asking to reverse translate the *Node  and its subtree back to original json shape. Thus this introduction of JSONify2 and its underlying J2NodeToInterface.